---
title: "AF_test"
output: html_document
date: "2023-03-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(qqman)
library(karyoploteR)
source("/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/likelihoodratiotest_function.R")
```

Write text file that only contains gnomad SNP id's
```{r}
gnomad_AC <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/gnomad/gnomad-genome_ssns_control.txt", sep = "\t", stringsAsFactors=FALSE, quote = "", header = TRUE)
gnomad_snp = gnomad_AC %>% mutate(SNP = paste0(CHROM, ":", POS, ":",  REF, ":", ALT)) %>% select(SNP)
# write.table(gnomad_snp,
#             "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_new/gnomad_snp_id.txt", row.names=FALSE, quote=FALSE, col.names = FALSE)
```

Merge discovery AC and gnomad AC for control vs control AF test
```{r}
discovery_AC <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_new/discovery_AC_ssns_allage.txt", header = TRUE, sep = " ")
AC_df = merge(gnomad_AC, discovery_AC, by = c("CHROM", "POS", "REF", "ALT"))
```

# AF test for all ancestry combined (df = 3)
```{r}
# define function input
# african
x1 = AC_df$disc_AC_af
n1 = AC_df$disc_AN_af
x2 = AC_df$AC_afr
n2 = AC_df$AN_afr
#south asian
x3 = AC_df$disc_AC_sas
n3 = AC_df$disc_AN_sas
x4 = AC_df$AC_sas
n4 = AC_df$AN_sas
#european
x5 = AC_df$disc_AC_euro
n5 = AC_df$disc_AN_euro
x6 = AC_df$AC_nfe
n6 = AC_df$AN_nfe

output_p = likelihoodratio_3(x1, n1, x2, n2, x3, n3, x4, n4, x5, n5, x6, n6) 

control_test = cbind(AC_df, output_p) %>% arrange(output_p) %>% 
  dplyr::rename(CHR = CHROM, BP = POS, P = output_p) %>% mutate(CHR = as.numeric(str_remove(CHR, "chr")), SNP = paste0(CHR, ":", BP)) %>% 
  arrange(CHR, BP)

# write.table(control_test,
#             "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_new/control_vs_control_pvals.txt", row.names=FALSE, quote=FALSE, col.names = TRUE)

list_to_remove = cbind(AC_df, output_p) %>% arrange(output_p) %>% filter(output_p < 6.797172e-06) 
list_to_remove

AC_df = AC_df %>% filter(!(CHROM %in% list_to_remove$CHROM & POS %in% list_to_remove$POS & REF %in% list_to_remove$REF & ALT %in% list_to_remove$ALT))
```

# get AC from bristol
```{r}
library(rio)
# reading data from all sheets
data <- import_list("../PHENOTYPE_BRISTOL_REPLICATION PLANS.xlsx") 
SSNS_sheet = data$SSNS[,1:5]
SSNS_age = data$ALL[,1:5] %>% mutate(AGE = as.numeric(AGE)) %>%  filter( DIAGNOSIS == "SSNS")

# write subset id's with phenotype
pheno_ <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/admixture/pca1_pheno_df.txt", sep = "\t", header = TRUE) %>% 
  select(IID, Sex, RACE, AGE, DIAGNOSIS, assign)
rownames(pheno_) <- NULL
table(pheno_$assign)
outlier1 = read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/admixture/outlier_labels1.txt")$V1
pheno_data = pheno_ %>% filter(! IID %in% outlier1) %>% select(IID, assign)
table(pheno_data$assign)
# merge with Rasheed's list
ssns_pheno_data = merge(pheno_data, SSNS_age, by.x = "IID", by.y = "AcquisitionNumber", all.y = TRUE) %>% drop_na(assign)
table(ssns_pheno_data$assign)
# write each ancestry into txt files
write.table(cbind(0, ssns_pheno_data %>% filter(assign == 1) %>% pull(IID)),
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/bristol_ssns_euro_allage.txt", row.names=FALSE, quote=FALSE, col.names = FALSE)
write.table(cbind(0, ssns_pheno_data %>% filter(assign == 2) %>% pull(IID)),
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/bristol_ssns_sas_allage.txt", row.names=FALSE, quote=FALSE, col.names = FALSE)
write.table(cbind(0, ssns_pheno_data %>% filter(assign == 3) %>% pull(IID)),
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/bristol_ssns_af_allage.txt", row.names=FALSE, quote=FALSE, col.names = FALSE)
write.table(cbind(0, ssns_pheno_data %>% pull(IID)),
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/bristol_ssns_allage.txt", row.names=FALSE, quote=FALSE, col.names = FALSE)

# some ref/alt mismatch due to multiallelic cases
# can now run LRT on Bristol cases vs Gnomad controls
# also perform LD clump on Bristol genotype data + LRT p-val outputs
```

Bristol case vs gnomad control AF test
```{r}
control_snp = AC_df %>% mutate(SNP = paste0(CHROM, ":", POS, ":",  REF, ":", ALT)) %>% select(SNP)
# write.table(control_snp,
#             "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_new/control_snp_id.txt", row.names=FALSE, quote=FALSE, col.names = FALSE)

bristol_AC <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_new/bristol_AC_ssns_allage.txt", header = TRUE) 
cases_vs_control = merge(gnomad_AC, bristol_AC, by = c("CHROM", "POS", "REF", "ALT"))
```


# LRT test on bristol
```{r}
# define function input
# african
x1 = cases_vs_control$bristol_AC_af
n1 = cases_vs_control$bristol_AN_af
x2 = cases_vs_control$AC_afr
n2 = cases_vs_control$AN_afr
#south asian
x3 = cases_vs_control$bristol_AC_sas
n3 = cases_vs_control$bristol_AN_sas
x4 = cases_vs_control$AC_sas
n4 = cases_vs_control$AN_sas
#european
x5 = cases_vs_control$bristol_AC_euro
n5 = cases_vs_control$bristol_AN_euro
x6 = cases_vs_control$AC_nfe
n6 = cases_vs_control$AN_nfe

output_p = likelihoodratio_3(x1, n1, x2, n2, x3, n3, x4, n4, x5, n5, x6, n6) 

# bonferroni correction:
nrow(cbind(cases_vs_control, output_p) %>% arrange(output_p) %>% filter(output_p < 8.058018e-06))

### write file for LD clump (All bristol snps)
ssns_ctr_full = cbind(cases_vs_control, output_p) %>% arrange(output_p) %>%
  mutate(SNP = paste0(CHROM, ":", POS, ":", REF, ":", ALT)) %>% select(SNP, P = output_p)
write.table(ssns_ctr_full, 
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_new/replication_unclump_ssns_allage.txt", 
            row.names=FALSE, quote=FALSE, col.names = TRUE, sep = "\t")

### clump Bristol p-values to get effective number of tests
clump_ssns_ctr_full <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_new/clump_bristol_6205.clumped", header = TRUE) 
nrow(clump_ssns_ctr_full)

# apply new p-value threshold
ssns_test = cbind(cases_vs_control, output_p) %>% arrange(output_p) %>% filter(output_p < 0.05/nrow(clump_ssns_ctr_full)) %>% 
  mutate(SNP = paste0(CHROM, ":", POS, ":", REF, ":", ALT))
length(intersect(clump_ssns_ctr_full$SNP, ssns_test$SNP))
intersect_list = intersect(clump_ssns_ctr_full$SNP, ssns_test$SNP)
ssns_test_66 = ssns_test %>% filter(SNP %in% intersect_list) %>% select(CHROM, POS, REF, ALT, ID, SNP, output_p)

write.table(ssns_test_66, 
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_new/bristol_ind_region_sig66.txt", 
            row.names=FALSE, quote=FALSE, col.names = TRUE, sep = "\t")
```


```{r}
# create input for karyoploteR package
ssns_ctr_full = cbind(cases_vs_control, output_p) %>% arrange(output_p) %>%
  mutate(SNP = paste0(CHROM, ":", POS, ":", REF, ":", ALT)) 
GR_input = ssns_ctr_full %>% mutate(chr = CHROM, start = POS, end = POS) %>% select(chr, start, end, A1 = REF, A2 = ALT, P = output_p)
ssns_gr = makeGRangesFromDataFrame(data.frame(GR_input), keep.extra.columns = TRUE)

ancestry_all = read.table("/datacommons/ochoalab/ssns_gwas/GMMAT_0418/ssns_ctr/glmm.score.bed.all_ssns_ctr_mac20_PC_nohwe.txt", sep = "\t", stringsAsFactors=FALSE, quote = "", header = TRUE)
results_all = ancestry_all %>% select(SNP, CHR, BP = POS, P = PVAL, A1, A2) %>% drop_na(P)

GR_input = results_all %>% mutate(chr = paste0("chr", CHR), start = BP, end = BP) %>% select(chr, start, end, A1, A2, P)
all_gr = makeGRangesFromDataFrame(data.frame(GR_input), keep.extra.columns = TRUE)
```



```{r fig.width=10, fig.height=5}
chr_plot = c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr11", 
             "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chr20", "chr21", "chr22")

pp <- getDefaultPlotParams(plot.type=4)
pp$ideogramlateralmargin = 0 
pp$ideogramheight = 0
pp$data1inmargin = -7

plot.new()
kp <- plotKaryotype("hg38", plot.type=4, ideogram.plotter = NULL, 
                    chromosomes = chr_plot, main = "Replication Analysis: SSNS vs Control", cex=1.5, plot.params = pp)

kpPlotManhattan(kp, data = ssns_gr, points.col="2blues", r0=autotrack(1,2, margin = 0.15),
                genomewide.col = "blue", genomewideline = -log10(1.72e-05), suggestive.col = "red", suggestiveline = -log10(0.000237))

kpAddLabels(kp, labels = "Bristol vs Gnomad", srt=90, pos=3, r0=autotrack(1,2, margin = 0.15), cex=1.5, label.margin = 0.035)
kpAxis(kp, ymin=0, ymax=20, r0=autotrack(1,2, , margin = 0.15), tick.pos = c(0,5,10,15,20))

kpPlotManhattan(kp, data = all_gr, points.col="2blues", r0=autotrack(2,2, margin = 0.15), , genomewideline = -log10(5e-8),
                suggestiveline = -log10(1e-5), , suggestive.col = "blue",
                genomewide.col = "red")
kpAddLabels(kp, labels = "Discovery set", srt=90, pos=3, r0=autotrack(2,2, margin = 0.15), cex=1.5, label.margin = 0.035)
kpAxis(kp, ymin=0, ymax=45, r0=autotrack(2,2, margin = 0.15))

```




