---
title: "AF_test"
output: html_document
date: "2023-03-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(qqman)
library(karyoploteR)
source("/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/likelihoodratiotest_function.R")
```

```{r}
# #gnomad data
# ssns_ctr_fulldata_gnomad <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/gnomad-3-genome.txt", header = TRUE, sep = "\t")
# 
# gnomad_df_id = ssns_ctr_fulldata_gnomad %>% mutate(SNP = paste0(CHROM, ":", POS, ":", REF, ":", ALT)) %>% select(SNP)
# write.table(gnomad_df_id,
#             "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/gnomad_snp_id.txt", row.names=FALSE, quote=FALSE, col.names = FALSE)
# 
# #bristol data
# ssns_ctr_fulldata_NS <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/ssns_control_all_AC.txt", header = TRUE, sep = "\t")
# 
# #combine gnomad & bristol
# AC_df = merge(ssns_ctr_fulldata_NS, ssns_ctr_fulldata_gnomad, by = c("CHROM", "POS", "REF", "ALT"), all.y = TRUE) %>% drop_na()
# write.table(AC_df, 
#             "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/NS_gnomad_ssns_all_AC.txt", 
#             sep = "\t", 
#             row.names=FALSE, quote=FALSE, col.names = TRUE)
#combined df
AC_df <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/NS_gnomad_ssns_all_AC.txt", header = TRUE, sep = "\t")
```

# AF test for all ancestry combined (df = 3)
```{r}
# define function input
#AC_df = AC_df %>% filter(!(CHROM %in% list_to_remove$CHROM & POS %in% list_to_remove$POS & REF %in% list_to_remove$REF & ALT %in% list_to_remove$ALT))
# african
x1 = AC_df$NS_AC_af
n1 = AC_df$NS_AN_af
x2 = AC_df$AC_afr
n2 = AC_df$AN_afr
#south asian
x3 = AC_df$NS_AC_sas
n3 = AC_df$NS_AN_sas
x4 = AC_df$AC_sas
n4 = AC_df$AN_sas
#european
x5 = AC_df$NS_AC_euro
n5 = AC_df$NS_AN_euro
x6 = AC_df$AC_nfe
n6 = AC_df$AN_nfe

output_p = likelihoodratio_3(x1, n1, x2, n2, x3, n3, x4, n4, x5, n5, x6, n6) 

# p-values for ssns p data
#output_all = likelihoodratio(x1, n1, x2, n2, 3)
hist(output_p, breaks = 50, main = "p-vals LRT < 1e-03; false positives removed ")
qq(output_p, main = "QQ-plot")

control_test = cbind(AC_df, output_p) %>% arrange(output_p) %>% 
  rename(CHR = CHROM, BP = POS, P = output_p) %>% mutate(CHR = as.numeric(str_remove(CHR, "chr")), SNP = paste0(CHR, ":", BP)) %>% 
  arrange(CHR, BP)

list_to_remove = cbind(AC_df, output_p) %>% arrange(output_p) %>% filter(output_p < 8.33e-06) 
list_to_remove
```


# LRT test on bristol
```{r}
bristol_AC <- read.table("bristol_ac_ancestry.txt", header = TRUE) 
b_g_AC = merge(bristol_AC, ssns_ctr_fulldata_gnomad, by = c("CHROM", "POS", "REF", "ALT"), all.y = TRUE) %>% drop_na()
write.table(b_g_AC, 
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/bristol_gnomad_ssns_all.txt", 
            row.names=FALSE, quote=FALSE, col.names = TRUE)

b_g_AC <- read.table("/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/bristol_gnomad_ssns_all.txt", header = TRUE, sep = " ")


subset_df = b_g_AC %>% filter(!(CHROM %in% list_to_remove$CHROM & POS %in% list_to_remove$POS & REF %in% list_to_remove$REF & ALT %in% list_to_remove$ALT))
# define function input
# african
x1 = subset_df$bristol_AF_af
n1 = subset_df$bristol_AN_af
x2 = subset_df$AC_afr
n2 = subset_df$AN_afr
#south asian
x3 = subset_df$bristol_AF_sas
n3 = subset_df$bristol_AN_sas
x4 = subset_df$AC_sas
n4 = subset_df$AN_sas
#european
x5 = subset_df$bristol_AF_euro
n5 = subset_df$bristol_AN_euro
x6 = subset_df$AC_nfe
n6 = subset_df$AN_nfe

output_p = likelihoodratio_3(x1, n1, x2, n2, x3, n3, x4, n4, x5, n5, x6, n6) 

# p-values for ssns p data
#output_all = likelihoodratio(x1, n1, x2, n2, 3)
hist(output_p, breaks = 50, main = "p-vals LRT < 2.075e-05; Bristol vs Gnomad control")
qq(output_p, main = "QQ-plot for all ancestry")

cbind(subset_df, output_p) %>% arrange(output_p) %>% filter(output_p < 0.000213) 

### write file for LD clump (All bristol snps)
ssns_ctr_full = cbind(subset_df, output_p) %>% arrange(output_p) 
  mutate(SNP = paste0(CHROM, ":", POS, ":", REF, ":", ALT)) 
write.table(ssns_ctr_full, 
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/replication_test_ssns_2913.txt", 
            row.names=FALSE, quote=FALSE, col.names = TRUE, sep = "\t")

### clump Bristol p-values to get effective number of tests
clump_ssns_ctr_full <- read.table("/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/newfiles_alpha/imputation_10_new/results/GMMAT/LD_clump/clump_bristol_all_r2_003_bp6_p1p2.txt.clumped", header = TRUE) 
nrow(clump_ssns_ctr_full)

### write sig snps into file for clumping
ssns_test = cbind(subset_df, output_p) %>% arrange(output_p) %>% filter(output_p < 0.000213) %>% 
  mutate(SNP = paste0(CHROM, ":", POS, ":", REF, ":", ALT)) %>% select(SNP, P = output_p)
write.table(ssns_test, 
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/replication_test_ssns_2913_1441sig.txt", 
            row.names=FALSE, quote=FALSE, col.names = TRUE, sep = "\t")
```


```{r}
# create input for karyoploteR package
GR_input = ssns_ctr_full %>% mutate(chr = CHROM, start = POS, end = POS) %>% select(chr, start, end, A1 = REF, A2 = ALT, P = output_p)
ssns_gr = makeGRangesFromDataFrame(data.frame(GR_input), keep.extra.columns = TRUE)

ancestry_all = read.table("/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/newfiles_alpha/imputation_10_new/results/GMMAT/ssns_ctr/full_data/glmm.score.bed.all_ssns_ctr_PC_20.txt", sep = "\t", stringsAsFactors=FALSE, quote = "", header = TRUE)
results_all = ancestry_all %>% select(SNP, CHR, BP = POS, P = PVAL, A1, A2) %>% drop_na(P)

GR_input = results_all %>% mutate(chr = paste0("chr", CHR), start = BP, end = BP) %>% select(chr, start, end, A1, A2, P)
all_gr = makeGRangesFromDataFrame(data.frame(GR_input), keep.extra.columns = TRUE)
```



```{r fig.width=10, fig.height=5}
chr_plot = c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr11", 
             "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chr20", "chr21", "chr22")

pp <- getDefaultPlotParams(plot.type=4)
pp$ideogramlateralmargin = 0 
pp$ideogramheight = 0
pp$data1inmargin = -7

plot.new()
kp <- plotKaryotype("hg38", plot.type=4, ideogram.plotter = NULL, 
                    chromosomes = chr_plot, main = "Replication Analysis", cex=1.5, plot.params = pp)

kpPlotManhattan(kp, data = ssns_gr, points.col="2blues", r0=autotrack(1,2, margin = 0.15),
                genomewide.col = "blue", genomewideline = -log10(3.86e-05), suggestive.col = "red", suggestiveline = -log10(0.000213))
#kpAddBase(h = -log10(2.074689e-05), col = "gray")
kpAddLabels(kp, labels = "Bristol vs Gnomad", srt=90, pos=3, r0=autotrack(1,2, margin = 0.15), cex=1.5, label.margin = 0.035)
kpAxis(kp, ymin=0, ymax=35, r0=autotrack(1,2, , margin = 0.15), tick.pos = c(0,5,10,15,20))

kpPlotManhattan(kp, data = all_gr, points.col="2blues", r0=autotrack(2,2, margin = 0.15), genomewideline = -log10(5e-8),
                suggestiveline = 7.3,
                genomewide.col = "red")
kpAddLabels(kp, labels = "SSNS vs Control", srt=90, pos=3, r0=autotrack(2,2, margin = 0.15), cex=1.5, label.margin = 0.035)
kpAxis(kp, ymin=0, ymax=50, r0=autotrack(2,2, margin = 0.15))

```

```{r}
### clump significant snps 1441
clump_1441 <- read.table("/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/newfiles_alpha/imputation_10_new/results/GMMAT/LD_clump/clump_bristol_all_r2_003_bp6_p1p2_1441sig.txt.clumped", header = TRUE) 

ssns_test_clumped = ssns_test %>% filter(SNP %in% clump_1441$SNP) %>% separate(SNP, c("chr", "pos", "A1", "A2")) %>% 
  mutate(chr = as.numeric(str_remove(chr, "chr")), pos = as.numeric(pos)) %>% dplyr::rename(LRT_P = P)
```


```{r}
ssns_sugsig_anno <- read.table("/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/newfiles_alpha/imputation_10_new/results/GMMAT/ssns_ctr/full_data/ssnsVScontrol_suggestivesig_rsid.txt", header = TRUE, sep = "\t") 

merged_sig_df = merge(ssns_test_clumped, ssns_sugsig_anno, by.x = c("chr", "pos", "A2", "A1"), by.y = c("Chromosome", "Position", "A1", "A2"), all.x = TRUE) %>% 
  arrange(LRT_P) %>% 
  dplyr::rename(replication_LRT_P_val = LRT_P) %>% select(chr, pos, A1, A2, rsid, original_AF = AF, original_SCORE = SCORE, original_VAR = VAR, original_PVAL = PVAL, everything(), -N)
length(unique(merged_sig_df$rsid))
write.table(merged_sig_df, 
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/replication_test_ssns_sig_80_annotated.txt", 
            row.names=FALSE, quote=FALSE, col.names = TRUE, sep = "\t")
```



```{r}
file = read.table("/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/newfiles_alpha/imputation_10_new/results/GMMAT/ssns_ctr/mac_20/SKAT/annovar/all_ssns_anno_step2_0215.hg38_multianno.txt", sep = "\t", header = TRUE)
merge2 = merge(merged_sig_df, file, by.x = c("chr", "pos", "A1", "A2"), by.y = c("Chr", "Start", "Ref", "Alt"), all.x = TRUE) %>% 
  arrange(replication_LRT_P_val) %>% select(-End, -Otherinfo1, -Otherinfo2, -Otherinfo3, -Otherinfo4, -Otherinfo5, -Otherinfo6, -Otherinfo7, -Otherinfo8, -Otherinfo9, -Otherinfo10, -Otherinfo11, -Otherinfo12)

write.table(merge2, 
            "/datacommons/ochoalab/ssns_gwas/replication/bristol_data/AF/ssns_ctr_all/replication_test_ssns_sig_80_annotated.txt", 
            row.names=FALSE, quote=FALSE, col.names = TRUE, sep = "\t")
```


