---
title: "susie_rpackage"
output: pdf_document
---

Susie analysis performed on the HLA region of chromosome 6
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("susieR")
library(tidyverse)
```

# x- genotypes
# y - cases/controls

#phenotypes and covariates
```{r}
phenodf <- read.table("/datacommons/ochoalab/ssns_gwas/phenotype_filtered.txt", sep = "\t", stringsAsFactors=FALSE, quote = "", header = TRUE) 

pheno_ = phenodf %>% mutate(Phenotype = ifelse(Phenotype == "Control", 0, 1)) 
```

```{r}
library(BEDMatrix)
#X <- BEDMatrix("/datacommons/ochoalab/tiffany_data/DATASOURCE/datasource/ssns_gwas_maf")
X <- BEDMatrix("/datacommons/ochoalab/tiffany_data/DATASOURCE/tmp_ld_ch6")
X <- as.matrix(X)
# X must be a double format
storage.mode(X) <- 'double'
rownames(X) <- NULL
colnames(X) <- NULL
# X cannot contain missing values
library(zoo)
X_agg = na.aggregate(X)
```


```{r}
fitted <- susie(X_agg, pheno_$Phenotype, L = 10, verbose = TRUE)
```

```{r}
print(fitted$sets)
```

```{r}
susie_plot(fitted, y="PIP", b=NULL)
```

```{r}
sumstats <- univariate_regression(X_agg, pheno_$Phenotype)
z_scores <- sumstats$betahat / sumstats$sebetahat
susie_plot(z_scores, y = "z", b=NULL)
```

```{r}
i  <- fitted$sets$cs[[3]]
z3 <- cbind(i,z_scores[i],fitted$pip[i])
colnames(z3) <- c('position', 'z-score', 'PIP')
susie_out = z3[order(z3[,2], decreasing = TRUE),] %>% as.data.frame()
snps = susie_out[,1]
```

```{r}
# read in bim file to get the chromosome information on significant SNPs. 
library(genio)
bim <- read_bim("/datacommons/ochoalab/tiffany_data/DATASOURCE/tmp_ld_ch6")
bim$position <- seq.int(nrow(bim))
bim[c(noquote(paste0(snps, collapse = ", "))), ]
bim_out = bim[c(773, 798, 794, 778, 786, 819, 790, 767, 784, 789, 800, 818, 820, 825, 859, 793, 837, 860), ]
```

```{r}
susie_output = merge(bim_out, susie_out, by = "position") %>% arrange(PIP)

```

