---
title: "allele_freq_preprocessing"
output: html_document
date: "2023-07-30"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(genio)
library(tidyverse)
library(ochoalabtools)
library(popgeninfer) # remotes::install_github('OchoaLab/popgeninfer')
library(scales) # for a color hack
```

## write tgp ancestry into text files
```{r}
# all the data is here
setwd( '/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/' )
fam <- read_fam( 'ssns_tgp_merge' )

# put outputs in a subdirectory
if ( !dir.exists( 'allele_freq' ) )
    dir.create( 'allele_freq' )
setwd( 'allele_freq' )

# define TGP ancestries
AFR <- c('MSL', 'YRI', "ESN", "GWD", "LWK", "ACB", "ASW")
EUR <- c('CEU', 'GBR', 'IBS', 'TSI', 'FIN')
SAS <- c('PJL', 'GIH', 'ITU', 'STU', 'BEB')

# save fam files with IDs for the desired individuals only
write_fam( "tgp_controls_eur", fam[ fam$fam %in% EUR, ] )
write_fam( "tgp_controls_sas", fam[ fam$fam %in% SAS, ] )
write_fam( "tgp_controls_afr", fam[ fam$fam %in% AFR, ] )
```

## use plink to calculate allele frequencies for controls from each ancestry

```bash
plink2 --bfile ../ssns_tgp_merge --keep ../admixture/ids_controls_afr.txt --freq --out array_controls_afr
plink2 --bfile ../ssns_tgp_merge --keep ../admixture/ids_controls_eur.txt --freq --out array_controls_eur
plink2 --bfile ../ssns_tgp_merge --keep ../admixture/ids_controls_sas.txt --freq --out array_controls_sas

plink2 --bfile ../ssns_tgp_merge --keep tgp_controls_afr.fam --freq --out tgp_controls_afr
plink2 --bfile ../ssns_tgp_merge --keep tgp_controls_eur.fam --freq --out tgp_controls_eur
plink2 --bfile ../ssns_tgp_merge --keep tgp_controls_sas.fam --freq --out tgp_controls_sas
```

## Read plink output files

Note the below tables are all aligned (each row is the same SNP in all of them) by construction, because they came from the same data (`ssns_tgp_merge`) without SNP filters.  The first four columns (CHROM, ID, REF, ALT) are identical in all files!

```{r}
# rest of processing happens here
setwd( '/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/' )
arr_afr <- read_tsv("array_controls_afr.afreq", col_types = 'icccdi')
arr_eur <- read_tsv("array_controls_eur.afreq", col_types = 'icccdi')
arr_sas <- read_tsv("array_controls_sas.afreq", col_types = 'icccdi')
tgp_afr <- read_tsv("tgp_controls_afr.afreq", col_types = 'icccdi')
tgp_eur <- read_tsv("tgp_controls_eur.afreq", col_types = 'icccdi')
tgp_sas <- read_tsv("tgp_controls_sas.afreq", col_types = 'icccdi')
```

# AF test for all ancestry combined (df = 3)

```{r}
# p-value threshold to use across analysis
pcut <- 1e-10

# recover allele counts from frequencies and sample sizes
arr_afr = arr_afr %>% mutate(AC = as.integer(round(ALT_FREQS * OBS_CT)))
tgp_afr = tgp_afr %>% mutate(AC = as.integer(round(ALT_FREQS * OBS_CT)))
arr_eur = arr_eur %>% mutate(AC = as.integer(round(ALT_FREQS * OBS_CT)))
tgp_eur = tgp_eur %>% mutate(AC = as.integer(round(ALT_FREQS * OBS_CT)))
arr_sas = arr_sas %>% mutate(AC = as.integer(round(ALT_FREQS * OBS_CT)))
tgp_sas = tgp_sas %>% mutate(AC = as.integer(round(ALT_FREQS * OBS_CT)))

# define function input
x1 <- cbind( arr_afr$AC, arr_eur$AC, arr_sas$AC )
n1 <- cbind( arr_afr$OBS_CT, arr_eur$OBS_CT, arr_sas$OBS_CT )
x2 <- cbind( tgp_afr$AC, tgp_eur$AC, tgp_sas$AC )
n2 <- cbind( tgp_afr$OBS_CT, tgp_eur$OBS_CT, tgp_sas$OBS_CT )
# p-values for forward alignment (only alignment for non-revcomp SNPs)
pvals <- af_test( x1, n1, x2, n2 )$pval
```

## Evaluate reverse complement ref/alt alleles

```{r}
# since all are SNPs, complements are reverse complements (i.e. no need to reverse)
comp <- function(x) chartr("ATGC","TACG",x)
# the "comp" rows are the same in all tables! (all datasets and ancestries) because they are aligned by construction
indexes_comp <- arr_afr$REF == comp( arr_afr$ALT )
# total starting SNPs
nrow( arr_afr ) # 762629
# get counts, are used later
m_comp <- sum( indexes_comp )
m_comp # 83708
m_noncomp <- sum( !indexes_comp )
m_noncomp # 678921
# get these IDs
ids_comp <- arr_afr$ID[ indexes_comp ]
# get subset of "forward" p-values for "comp" cases
pvals_comp_fwd <- pvals[ indexes_comp ]

# now calculate "reverse" orientation p-values, for "comp" cases only
# subset to comp SNPs
x1c <- x1[ indexes_comp, ]
n1c <- n1[ indexes_comp, ]
x2c <- x2[ indexes_comp, ]
n2c <- n2[ indexes_comp, ]
# perform reversed test, reverse first dataset only (SSNS), not second (TGP)
pvals_comp_rev <- af_test( n1c - x1c, n1c, x2c, n2c )$pval

# classify SNPs based in their p-values (the same for all ancestries!), using this decision tree
category_comp <- ifelse( pvals_comp_fwd < pcut & pvals_comp_rev < pcut, "Removed",
                 ifelse( pvals_comp_fwd < pvals_comp_rev, "Flipped", "Unchanged" ) )
# set plotting order
category_comp <- factor(category_comp, levels = c("Unchanged", "Flipped", "Removed"))
# inspect
as_tibble( table( category_comp ) )
##   category_comp     n
## 1 Unchanged     80310
## 2 Flipped        3178
## 3 Removed         220
# counts when we were counting both_nonsig_{keep,flip} cases separately, for reference:
##   category_comp        n
## 1 Unchanged        80037
## 2 Flipped           3133
## 3 Removed            220
## 4 both_notsig_keep   273
## 5 both_notsig_flip    45
# lists of "comp" IDs to flip or remove
ids_comp_remove <- ids_comp[ category_comp == "Removed" ]
ids_comp_flip <- ids_comp[ category_comp == "Flipped" ]

# to prep for big plot later, arrange all data in a single data frame
# make it easy to facet by ancestry and "comp" status
# includes p-values for a separate plot, and IDs just for reference (not used by plots)
data_comp <- tibble(
    ids = rep.int( ids_comp, 3 ),
    WGS   = c( tgp_afr$ALT_FREQS[ indexes_comp ], tgp_eur$ALT_FREQS[ indexes_comp ], tgp_sas$ALT_FREQS[ indexes_comp ] ),
    Array = c( arr_afr$ALT_FREQS[ indexes_comp ], arr_eur$ALT_FREQS[ indexes_comp ], arr_sas$ALT_FREQS[ indexes_comp ] ),
    Category = rep.int( category_comp, 3 ),
    ancestry = c( rep.int( 'African', m_comp ), rep.int( 'European', m_comp ), rep.int( 'South Asian', m_comp ) ),
    comp = 'Reverse complement ref/alt',
    pvals_fwd = rep.int( pvals_comp_fwd, 3 ),
    pvals_rev = rep.int( pvals_comp_rev, 3 )
)
```

## Evaluate non-reverse complement ref/alt alleles

```{r}
# p-values for non-comp SNPs
pvals_noncomp <- pvals[ !indexes_comp ]
# subset datasets too, overwriting original tables
tgp_afr <- tgp_afr[ !indexes_comp, ]
tgp_eur <- tgp_eur[ !indexes_comp, ]
tgp_sas <- tgp_sas[ !indexes_comp, ]
arr_afr <- arr_afr[ !indexes_comp, ]
arr_eur <- arr_eur[ !indexes_comp, ]
arr_sas <- arr_sas[ !indexes_comp, ]

# classify SNPs (same for all ancestries)
category_noncomp <- ifelse( pvals_noncomp < pcut , "Removed", "Unchanged")
# set order for plotting
category_noncomp <- factor(category_noncomp, levels = c("Unchanged", "Removed"))
# get counts
as_tibble( table( category_noncomp ) )
##   category_noncomp      n
## 1 Unchanged        677878
## 2 Removed            1043

# get IDs to remove based on this analysis
ids_noncomp_remove <- tgp_afr$ID[ category_noncomp == "Removed" ]

# prep data for big plot, as before
data_noncomp <- tibble(
    ids = rep.int( tgp_afr$ID, 3 ),
    WGS   = c( tgp_afr$ALT_FREQS, tgp_eur$ALT_FREQS, tgp_sas$ALT_FREQS ),
    Array = c( arr_afr$ALT_FREQS, arr_eur$ALT_FREQS, arr_sas$ALT_FREQS ),
    Category = rep.int( category_noncomp, 3 ),
    ancestry = c( rep.int( 'African', m_noncomp ), rep.int( 'European', m_noncomp ), rep.int( 'South Asian', m_noncomp ) ),
    comp = 'Non-reverse complement ref/alt',
    pvals_fwd = rep.int( pvals_noncomp, 3 ),
    pvals_rev = NA
)
```

## Write SNP id's to remove/flip into txt file

```{r}
# same as originals, but order is different
write_lines( c( ids_comp_remove, ids_noncomp_remove ), "remove.txt")
write_lines( ids_comp_flip, "flip.txt")
```

## Save plotting data

Save this big tibble used to make plots, to just load back if we're just tweaking plots and analysis didn't change

```{r}
# combine all data to plot into a single tibble
data <- bind_rows( data_comp, data_noncomp )
# save!
write_tsv( data, "af-test.txt.gz")
```

## Reload plotting data if needed

Skip if analysis was redone from the start and everything is loaded in the session already.

Run if this is a new session and we're only tweaking plots, to reload plotting data only.

```{r}
# Reload these libraries too
library(tidyverse)
library(ochoalabtools)
library(scales) # for a color hack
# and make sure we're in the correct path.
setwd( '/datacommons/ochoalab/ssns_gwas/nephrotic.syndrome.gwas_proprocessing_202205/allele_freq/' )
# load big tibble
data <- read_tsv( "af-test.txt.gz", col_types = 'cddcccdd' )
# reset plotting order (not preserved in flattened tibble)
data$Category <- factor( data$Category, levels = c("Unchanged", "Flipped", "Removed") )

# if making p-value plot, extract that data from the tibble this way:
# first subset to a single ancestry (which one doesn't matter, p-values are the same for each, but they are repeated otherwise)
data_afr <- filter( data, ancestry == 'African' )
# reidentify comp cases
indexes_comp2 <- data_afr$comp == 'Reverse complement ref/alt'
# now recover p-value distributions
pvals_noncomp <- data_afr$pvals_fwd[ !indexes_comp2 ]
pvals_comp_fwd <- data_afr$pvals_fwd[ indexes_comp2 ]
pvals_comp_rev <- data_afr$pvals_rev[ indexes_comp2 ]
```


## P-value distribution plots

Make a figure with all cases together as panels.

```{r}
# these are full-page figs, scale (w/h) appropriately
dims <- fig_scale( 1 )
fig_start(
    'pvals',
    width = dims[1],
    height = dims[2],
    mar_t = 1.5
)
# 3 rows, one column
par( mfrow = c(3, 1) )
hist( pvals_noncomp, freq = FALSE, breaks = 50, main = "Non-reverse complement ref/alt", xlab = '' )
abline( h = 1, lty = 2, col = 'gray' )
hist( pvals_comp_fwd, freq = FALSE, breaks = 50, main = "Reverse complement ref/alt: forward orientation", xlab = '' )
abline( h = 1, lty = 2, col = 'gray' )
hist( pvals_comp_rev, freq = FALSE, breaks = 50, main = "Reverse complement ref/alt: reverse orientation", xlab = 'p-value' )
abline( h = 1, lty = 2, col = 'gray' )
fig_end()
```

## Allele frequency and SNP classification plots

Combine all data into a nice plot faceted by ancestry and "comp" status:

```{r}
# this plots "Unchanged" first (at back), followed by "Flipped", "Removed", in order of Category factor, so we can see smaller categories most clearly
data <- data %>% arrange( Category )

# get default colors for these many categories, but reorder
colors <- hue_pal()(3) # default (red, green, blue)
colors <- c( "#999999", colors[ c(3, 1) ] ) # want (Unchanged=darkgray, Flipped=blue, Removed=red )

# visualize results by ancestry
# these are full-page figs, scale (w/h) appropriately
dims <- fig_scale( 1 )
fig_start(
    'af-test',
    width = dims[1], # 7.5 for scale=1
    height = dims[2] # 7.5 for scale=1
)
ggplot( data, aes( x = Array, y = WGS, color = Category ) ) +
    labs( x = 'Array allele frequency', y = 'WGS allele frequency' ) +
    theme_classic() +
    xlim( 0, 1 ) +
    ylim( 0, 1 ) +
    geom_point( size = 0.25 ) + 
    scale_color_manual( values = colors ) +
    guides( colour = guide_legend( override.aes = list( size = 1 ) ) ) + # increase point size in legend only
    facet_grid( ancestry ~ comp )
fig_end()
```
